feat: Add CLP connector

wraymo opened the pull request:
## Overview
The current Presto–CLP connector PR introduces the coordinator-side implementation, along with a placeholder (dummy) worker implementation. Detailed information about the overall design is available in the corresponding RFC. This Velox PR focuses on the worker-side logic.

The Velox CLP connector enables query execution on CLP archives. The Velox worker receives split information and the associated KQL query from the Presto coordinator. For each split, it executes the KQL query against the relevant CLP archive to find matching messages and stores their indices.

To support lazy evaluation, the implementation creates lazy vectors that wrap a CLP column reader and the list of matching indices. When accessed during query execution, these vectors load and decode only the necessary data on demand.

## Core Classes
### `ClpDataSource`
This class extends `DataSource` and implements the `addSplit` and `next` methods. During initialization, it records the KQL query and archive source (S3 or local), then traverses the output type to map Presto fields to CLP projection fields. Only `ARRAY(VARCHAR)` and primitive leaf fields like `BIGINT`, `DOUBLE`, `BOOLEAN` and `VARCHAR` are projected.

When a split is added, a `ClpCursor` is created with the archive path and input source. The query is parsed and simplified into an AST. On `next`, the cursor finds matching row indices and, if any exist, returns a row vector composed of lazy vectors, which load data as needed during execution.

### `ClpCursor`
This class manages the execution of a query over a CLP-S archive. It handles parsing and validation, loading schemas and archives, setting up projection fields, and filtering results. In CLP-S, records are partitioned by schemas. `ClpCursor` uses `ClpQueryRunner` to initialize the execution context for each schema and evaluate the filters. It will skip archives where dictionary lookups for string filters return no matches and only scan the relevant schemas of a specific archive. For example, consider a log dataset with the following records.
```
{"a": "Hello", "b": 2}
{"a": "World", "b": 0, "c": false}
{"a": "World", "c": true}
```
The three log messages have varying schemas. If we run a KQL query `a: World AND b: 0`, it will skip loading the third message because it's schema does not match the query (there's no `b` field). And if the query is `a: random AND b: 0`, it will even skip scanning the first two records, because `random` cannot be found in the dictionary.

### `ClpQueryRunner`
This class extends the generic CLP `QueryRunner` to support ordered projection and row filtering. It initializes projected column readers and returns filtered row indices for each batch.

### `ClpVectorLoader`
In CLP, values are decoded and read from a `BaseColumnReader`. The `ClpVectorLoader` is custom Velox `VectorLoader` that loads vectors from CLP column readers. It supports integers, floats, booleans, strings, and arrays of strings. It's used by lazy vectors to load data on demand using the previously stored row indices.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced the CLP Connector for reading CLP archives from local filesystems and S3 with support for KQL query pushdown.
  - Added lazy loading and filtering of data with support for nested row types and complex query expressions.
  - Integrated build support for the CLP Connector, including new dependencies such as spdlog, msgpack-cxx, and ANTLR.
  - Provided a new search library and query engine to power CLP data access and filtering.
- **Tests**
  - Added comprehensive unit tests and example NDJSON datasets to validate CLP Connector functionality and query correctness.
- **Documentation**
  - Updated Velox connector documentation to include detailed descriptions and usage examples for the CLP Connector.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

wraymo committed:
add clp module

wraymo committed:
add clp code to velox

wraymo committed:
add more methods

wraymo committed:
add more methods

wraymo committed:
modify CMakeLists.txt

wraymo committed:
add clp data source support

wraymo committed:
add polymorphic type support

wraymo committed:
remove duplicate values and add null support

wraymo committed:
fix array bugs and the pipe blocking bug

wraymo committed:
add concurrent split support

wraymo committed:
fixed a bug affecting select count(*) queries

wraymo committed:
apply changes from upstream

wraymo committed:
fix some issues

wraymo committed:
fix some bugs for concurrent ClppSplits

wraymo committed:
fix some issues

wraymo committed:
another impl with fork and pipe

wraymo committed:
add clp lib code for clp connector

wraymo committed:
solve many dependency issues

wraymo committed:
add antlr jar

wraymo committed:
add single file archive and network reader support

wraymo committed:
add msgpack dependency

wraymo committed:
add datasouce url support

wraymo committed:
add s3 support

wraymo committed:
fix several issues

wraymo committed:
not treat warnings as errors

wraymo committed:
change the order

wraymo committed:
refactor the code

wraymo committed:
update clp dep

wraymo committed:
update velox code

wraymo committed:
rebase to the latest master branch

wraymo committed:
revert IP related changes

Commit:
update connector APIs and change the way how we use clpsplit

Commit:
fix warnings

wraymo committed:
advance clp

wraymo committed:
working on QueryRunner refactoring

wraymo committed:
refactor the code

wraymo committed:
finish refactoring

wraymo commented on third-party/antlr/antlr-4.13.1-complete.jar:
We may need to come up with a better way to add the dependency

anlowee replied:
Is this issue resolved?

wraymo committed:
add license header

wraymo committed:
fix lint errors in cmake files

wraymo committed:
improve clp connector comments

wraymo committed:
improve clp connector comments and reflect clp src code changes in clp connector

wraymo committed:
add support to track num of rows sacnned

wraymo committed:
download antlr jar on demand and clean up FindANTLR.cmake

wraymo committed:
add ClpConnectorTest

wraymo committed:
use loop-based for

wraymo committed:
advance clp submodule

wraymo committed:
add clp test changes

anlowee commented on Makefile:
```diff
@@ -19,7 +19,7 @@ BUILD_DIR=release
 BENCHMARKS_BASIC_DIR=$(BUILD_BASE_DIR)/$(BUILD_DIR)/velox/benchmarks/basic/
 BENCHMARKS_DUMP_DIR=dumps
-TREAT_WARNINGS_AS_ERRORS ?= 1
+TREAT_WARNINGS_AS_ERRORS = 0
```
I think this could only be a debugging twist.

anlowee commented on velox/connectors/clp/ClpDataSource.cpp:
```diff
@@ -0,0 +1,183 @@
+    std::shared_ptr<const ClpConfig>& clpConfig)
+    : pool_(pool), outputType_(outputType) {
+  archiveSource_ = clpConfig->archiveSource();
+  boost::algorithm::to_lower(archiveSource_);
```
why?

wraymo replied:
Because it might be upper case in the config option. See https://github.com/facebookincubator/velox/blob/main/velox/connectors/hive/HiveConfig.cpp#L29

anlowee commented on velox/connectors/clp/ClpDataSource.cpp:
```diff
@@ -0,0 +1,183 @@
+  archiveSource_ = clpConfig->archiveSource();
+  boost::algorithm::to_lower(archiveSource_);
+  auto clpTableHandle = std::dynamic_pointer_cast<ClpTableHandle>(tableHandle);
+  if ("local" != archiveSource_ && "s3" != archiveSource_) {
```
Can these two be two lines similar to this `static constexpr const char* kArchiveSource = "clp.archive-source";` in `ClpConfig`?

wraymo replied:
what do you mean

wraymo replied:
I think we can make them enum, similar to https://github.com/facebookincubator/velox/blob/main/velox/connectors/hive/HiveConfig.cpp#L29

anlowee commented on velox/connectors/clp/ClpDataSource.cpp:
```diff
@@ -0,0 +1,183 @@
+  if (query && !query->empty()) {
+    kqlQuery_ = *query;
+  } else {
+    kqlQuery_ = "*";
```
shall we throw error rather than transiciently replace it with *? I think this should be an USER_ERROR?

wraymo replied:
`query` can be empty if we cannot pushdown any filters. It will be translated to `*` in kql. But since we are changing `query` to `kqlQuery` in Presto, we also need to update the protocol code and probably rename it in relevant classes .

anlowee replied:
Got it. If you are going to rename it in Presto PR, then let's rename it in this PR as well as presto-native PR.

anlowee commented on velox/connectors/clp/ClpDataSource.cpp:
```diff
@@ -0,0 +1,183 @@
+
+  if (archiveSource_ == "local") {
+    cursor_ = std::make_unique<search_lib::ClpCursor>(
+        clp_s::InputSource::Filesystem, clpSplit->archivePath_);
```
As suggested by Jack, we can rename all `archivePath*` to `splitPath`, because it confuses when we have IR files and SFA (where only the latter one is "archive")

anlowee commented on velox/connectors/clp/ClpDataSource.cpp:
```diff
@@ -0,0 +1,183 @@
+  cursor_->executeQuery(kqlQuery_, fields_);
+}
+
+VectorPtr ClpDataSource::createVector(
```
shall we add some comments to briefly example what we are doing here?

anlowee commented on velox/connectors/clp/search_lib/ClpCursor.cpp:
```diff
@@ -0,0 +1,236 @@
+      archivePath_(std::move(archivePath)) {}
+
+ErrorCode ClpCursor::loadArchive() {
+  auto networkAuthOption = inputSource_ == InputSource::Filesystem
+      ? NetworkAuthOption{.method = AuthMethod::None}
+      : NetworkAuthOption{.method = AuthMethod::S3PresignedUrlV4};
```
Do we want to support IR search in this PR as well

wraymo replied:
Not sure. @kirkrodrigues Do you think we should support it in the first PR?

kirkrodrigues replied:
No, let's leave that to a future PR since we won't have support for it in the clp-package yet.

anlowee commented on velox/connectors/clp/search_lib/ClpCursor.cpp:
```diff
@@ -0,0 +1,236 @@
+  auto schemaMatch = std::make_shared<SchemaMatch>(schemaTree, schemaMap);
+  if (expr_ = schemaMatch->run(expr_);
+      std::dynamic_pointer_cast<EmptyExpr>(expr_)) {
+    // SPDLOG_ERROR("No matching schemas for query '{}'", query_);
```
Remove the comment-out?

wraymo replied:
I'm more inclined to remove this line since there's no log printed in Velox

anlowee commented on velox/connectors/clp/search_lib/ClpCursor.cpp:
```diff
@@ -0,0 +1,236 @@
+      if (false ==
+          tokenize_column_descriptor(
+              column.name, descriptorTokens, descriptorNamespace)) {
+        // SPDLOG_ERROR("Can not tokenize invalid column: \"{}\"", column);
```
same

anlowee commented on velox/connectors/clp/search_lib/ClpCursor.cpp:
```diff
@@ -0,0 +1,236 @@
+      projection_->add_column(columnDescriptor);
+    }
+  } catch (TraceableException& e) {
+    SPDLOG_ERROR("{}", e.what());
```
I am thinking of do we want to use Velox's logger rather than SPDLOG? I remember this code is copied from CLP, but now since major part of CLP code is already in the submodule, so we should make the rest of code match Velox style.

anlowee replied:
Some of them are good, for example the one for `rowFecthed xx`, I always use it to check if the worker is working. So maybe just use what they are using (i.e., VLOG)

anlowee commented on velox/connectors/clp/search_lib/ClpCursor.cpp:
```diff
@@ -0,0 +1,236 @@
+  auto queryStream = std::istringstream(query_);
+  expr_ = kql::parse_kql_expression(queryStream);
+  if (nullptr == expr_) {
+    // SPDLOG_ERROR("Failed to parse query '{}'", query_);
```
Same, remove all comment-out pls.

anlowee commented on velox/connectors/clp/search_lib/CMakeLists.txt:
```diff
@@ -0,0 +1,202 @@
+                           PUBLIC SOURCE_PATH_SIZE=${SOURCE_PATH_SIZE})
+velox_include_directories(
+  clp-s-search PUBLIC ${ANTLR_KqlParser_OUTPUT_DIR}
+                      clp_src/components/core/submodules ${CLP_SRC_DIR})
```
add ystdlib?

wraymo replied:
Is it required?

anlowee commented on velox/connectors/clp/search_lib/CMakeLists.txt:
```diff
@@ -0,0 +1,202 @@
+    ${CLP_SRC_DIR}/clp_s/search/kql/kql.hpp
+    ${CLP_SRC_DIR}/clp_s/search/Projection.cpp
+    ${CLP_SRC_DIR}/clp_s/search/Projection.hpp
+    ${CLP_SRC_DIR}/clp_s/search/SchemaMatch.cpp
+    ${CLP_SRC_DIR}/clp_s/search/SchemaMatch.hpp
```
Also add QueryRunner.hpp/cpp, otherwise it won't build

wraymo commented on velox/connectors/clp/ClpConfig.h:
```diff
@@ -0,0 +1,47 @@
+namespace facebook::velox::connector::clp {
+class ClpConfig {
+ public:
+  static constexpr const char* kArchiveSource = "clp.archive-source";
```
Seems that hive declares constant variables in the same way.

wraymo committed:
resolve comments

wraymo committed:
remove spdlog in ClpCursor.cpp

wraymo committed:
add QueryRunner to CMakeLists.txt

wraymo committed:
add glog statements in ClpCursor.cpp

wraymo committed:
unit test progress

wraymo committed:
fix multiple bugs

wraymo committed:
optimize headers

wraymo committed:
add the second test case; change archivePath to splitPath

wraymo committed:
add more tests

wraymo committed:
fix code formats

wraymo committed:
add clp docs

anlowee commented on velox/connectors/clp/search_lib/ClpCursor.cpp:
```diff
@@ -0,0 +1,236 @@
+      inputSource_(inputSource),
+      archivePath_(std::move(archivePath)) {}
+
+ErrorCode ClpCursor::loadArchive() {
```
Do we want to address this as well? I see at other places we already renamed `archive*` to `split*`.

anlowee commented on CMake/resolve_dependency_modules/spdlog.cmake:
```diff
@@ -0,0 +1,35 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
```
If we have removed the SPDLOG (those comment-out) do we still need this? If not maybe also clean the related CMakeLists.txt?

wraymo replied:
We cannot remove it because some clp files need that even if velox doesn't need

anlowee replied:
I am a bit confused. I think those CLP files should be compiled within `clp_src` submodule which should have its own cmake rules including the spdlog? WDYT @kirkrodrigues

anlowee replied:
I will try compile without it from my end to see what will go wrong

wraymo replied:
If you take a look at the cmake file, it didn't use `add_subdirectory`, but directly include all source files

kirkrodrigues replied:
> clp_src submodule which should have its own cmake rules including the spdlog

Dependency-installation should be handled by the top-level project otherwise you could have conflicts where clp tries to install version X of spdlog and Velox tries to install version Y of spdlog.

If this is the correct way to add a new dependency to Velox, then we should leave this file.

wraymo replied:
> I will try compile without it from my end to see what will go wrong
@anlowee What's the result?

kirkrodrigues replied:
I would prefer that we still leave getting SPDLOG to the top-level project because:

* it's hard to predict what Velox will do in the future.
* if we force the CLP submodule to pull SPDLOG, then it also breaks other projects that depend on CLP.

anlowee commented on CMakeLists.txt:
```diff
@@ -485,6 +488,20 @@ endif()
 
+if(${VELOX_ENABLE_CLP_CONNECTOR})
+  set(spdlog_SOURCE BUNDLED)
+  velox_resolve_dependency(spdlog 1.12.0)
```
If no SPDLOG anymore, maybe remove it.

wraymo commented on velox/connectors/clp/ClpConnectorSplit.h:
```diff
@@ -0,0 +1,41 @@
+        splitPath_(splitPath) {}
+
+  [[nodiscard]] std::string toString() const override {
+    return fmt::format("CLP: {}.{}", schemaName_, tableName_);
```
Ok, and we can probably remove the schema name and the table name in the future

anlowee commented on velox/connectors/clp/ClpDataSource.cpp:
```diff
@@ -0,0 +1,176 @@
+    const std::string& parentName) {
+  if (columnType->kind() == TypeKind::ROW) {
+    const auto& rowType = columnType->asRow();
+    for (size_t i = 0; i < rowType.size(); ++i) {
```
i{0}

anlowee commented on velox/connectors/clp/ClpDataSource.cpp:
```diff
@@ -0,0 +1,176 @@
+    return std::make_shared<RowVector>(
+        pool_, type, nullptr, size, std::move(children));
+  }
+  auto vector = BaseVector::create(type, size, pool_);
+  vector->setNulls(allocateNulls(size, pool_, bits::kNull));
+  auto projectedColumn = projectedColumns[readerIndex];
+  auto projectedType = fields_[readerIndex].type;
+  readerIndex++;
+  return std::make_shared<LazyVector>(
+      pool_,
+      type,
+      size,
+      std::make_unique<search_lib::ClpVectorLoader>(
+          projectedColumn, projectedType, filteredRows),
+      std::move(vector));
```
I personally like write the recursion termination handling at the beginning of the recursive function, even though here is impossible to do tail call optimization.

anlowee commented on velox/connectors/clp/ClpDataSource.cpp:
```diff
@@ -0,0 +1,176 @@
+  auto rowsScanned = cursor_->fetch_next(size, filteredRows);
+  auto rowsFiltered = filteredRows->size();
+  if (rowsFiltered == 0) {
+    return nullptr;
```
std::nullopt

wraymo committed:
temperory solution

anlowee commented on velox/connectors/clp/search_lib/ClpCursor.cpp:
```diff
@@ -0,0 +1,255 @@
+      archiveReader_(std::make_shared<ArchiveReader>()) {}
+
+ClpCursor::~ClpCursor() {
+  if (currentArchiveLoaded_) {
+    archiveReader_->close();
```
archive -> split

wraymo replied:
It's ArchiveReader in CLP.

anlowee commented on velox/connectors/clp/search_lib/ClpCursor.h:
```diff
@@ -0,0 +1,124 @@
+   * query result.
+   * @return The error code.
+   */
+  ErrorCode executeQuery(
```
in `ClpDataSource` seems only place calls this function is this line: `cursor_->executeQuery(kqlQuery_, fields_);`, which doesn't care about the returned `ErrorCode`.

anlowee commented on velox/connectors/clp/search_lib/ClpCursor.h:
```diff
@@ -0,0 +1,124 @@
+
+namespace facebook::velox::connector::clp::search_lib {
+
+enum class ErrorCode {
+  Success,
+  QueryNotInitialized,
+  InvalidQuerySyntax,
+  SchemaNotFound,
+  LogicalError,
+  DictionaryNotFound,
+  InvalidTimestampRange,
+  InternalError
+};
```
Is this error code only for connector internal use (i.e., will not return error message to the user) or will eventually return to Presto as a PrestoException (i.e., user-aware)? If the later one, maybe we need to finally map these error code to those in `VeloxException.h`, otherwise it will just be some generic error shown to users (e.g., `GENERIC_INTERNAL_ERROR`, `GENERIC_USER_ERROR` etc.)

wraymo replied:
It's for internal use.

wraymo committed:
address comments

anlowee commented on velox/connectors/clp/search_lib/ClpVectorLoader.cpp:
```diff
@@ -0,0 +1,155 @@
+
+template <typename T, typename VectorPtr>
+void ClpVectorLoader::populateData(RowSet rows, VectorPtr vector) {
+  if (columnReader_ == nullptr) {
```
Maybe just `if (columnReader_)`

wraymo replied:
We want to handle this case first

anlowee commented on velox/connectors/clp/search_lib/ClpCursor.cpp:
```diff
@@ -0,0 +1,255 @@
+  auto queryStream = std::istringstream(query_);
+  expr_ = kql::parse_kql_expression(queryStream);
+  if (nullptr == expr_) {
+    DLOG(ERROR) << "Failed to parse query '" << query_ << "'";
```
Yea please replace

wraymo committed:
use fetch_content for clp

wraymo committed:
fix code formats issues

wraymo committed:
remove clp header files from CMakeLists.txt

wraymo committed:
change method name

wraymo commented on velox/connectors/clp/tests/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+add_executable(velox_clp_connector_test ClpConnectorTest.cpp)
```
What's this? Didn't find it

anlowee commented on CMake/ExternalAntlr4Cpp.cmake:
```diff
@@ -0,0 +1,175 @@
+      -DDISABLE_WARNINGS:BOOL=ON
+      # -DCMAKE_CXX_STANDARD:STRING=17 # if desired, compile the runtime with a
+      # different C++ standard -DCMAKE_CXX_STANDARD:STRING=${CMAKE_CXX_STANDARD}
+      # # alternatively, compile the runtime with the same C++ standard as the
```
extra `#`

anlowee commented on CMake/ExternalAntlr4Cpp.cmake:
```diff
@@ -0,0 +1,175 @@
+      -DDISABLE_WARNINGS:BOOL=ON
+      # -DCMAKE_CXX_STANDARD:STRING=17 # if desired, compile the runtime with a
+      # different C++ standard -DCMAKE_CXX_STANDARD:STRING=${CMAKE_CXX_STANDARD}
+      # # alternatively, compile the runtime with the same C++ standard as the
```
same

wraymo committed:
fix several build issues and optimize the cmake files

wraymo committed:
1. remove extra # in cmake files
2. fix a bug in FindANTLR.cmake
3. download antlr jar to build directory instead of the source directory
4. add StorageType to ClpTableHandle
5. use consistent types for filteredRows across different methods
6. remove archiveSource_ from ClpConfig and config docs, use storageType_ in ClpDataSource
7. minor improvements in logs and method signatures
8. rename splitPath to path in ClpConnectorSplit

anlowee commented on CMakeLists.txt:
```diff
@@ -485,6 +489,23 @@ endif()
+  velox_resolve_dependency(clp)
+
+  set(spdlog_SOURCE BUNDLED)
+  velox_resolve_dependency(spdlog 1.12.0)
```
you already hardcoded the version in `spdlog.cmake`, maybe get rid of 1.12.0 here

anlowee commented on velox/connectors/clp/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+  ClpDataSource.cpp
+  ClpTableHandle.cpp)
+
+velox_link_libraries(velox_clp_connector PRIVATE clp-s-search velox_connector
```
maybe switch the order of simdjson and velox_connector

anlowee commented on velox/connectors/clp/search_lib/ClpQueryRunner.cpp:
```diff
@@ -0,0 +1,82 @@
+      continue;
+    }
+
+    // Try to find a matching column in m_column_map
```
is this commnent outdate? I can't find `m_column_map`

anlowee commented on velox/connectors/clp/ClpDataSource.cpp:
```diff
@@ -0,0 +1,175 @@
+        break;
+      case TypeKind::INTEGER:
+      case TypeKind::BIGINT:
+      case TypeKind::TINYINT:
```
Why we don't support `SMALLINT`?

wraymo replied:
We can support, just according to our mapping, no type can be mapped to SMALLINT (also INTEGER)

wraymo committed:
1. rename varialbes in ClpQueryRunner
2. use consistent type for numRows in ClpQueryRunner
3. remove spdlog version in CMakeLists.txt and change find_package to resolve_dependency for curl
4. fix readerIndex type in ClpDataSource
5. add SMALLINT in ClpDataSource
